### ISSUE TITLE
Threadproducer is not throwing error evenif the server does not exist

### ISSUE THREAD

**Issue by wensiwei** [fede1024/rust-rdkafka#540]
I'm used [`ThreadedProducer`] to send kafka messages. I also hope to get the error information of send when the server does not exist, but I don't know how to catch this error in time . The message can't be lost.

So how do I get the producer result in the example,  thank you!

code
```rust
#[macro_use]
extern crate tracing;

use std::thread;
use std::time::Duration;
use rdkafka::ClientConfig;
use rdkafka::producer::{ThreadedProducer, BaseRecord, DefaultProducerContext};
use tlog::{init_global_log, LogConfig};

#[tokio::main]
async fn main() {
    let new_cfg = LogConfig::default().set_level("debug");
    let (_file_guard, _log_manager) = init_global_log(&new_cfg).unwrap();

    let producer: ThreadedProducer<DefaultProducerContext> = ClientConfig::new()
        .set("bootstrap.servers", "locast:9092")
        .create()
        .unwrap();

    println!("producer created");

    for i in 1..10{
        let message = format!("message {}", i);
        let s = producer
            .send(
                BaseRecord::to("quickstart-events")
                    .payload(&message)
                    .key("")
            );
        println!("{:?}", s);
        thread::sleep(Duration::from_secs(1));
    }
}
```

result
```linux
    Finished dev [unoptimized + debuginfo] target(s) in 6.22s
     Running `target/debug/test_thread_producer`
producer created
Ok(())
  2023-01-29T17:06:11.284 ERROR librdkafka: librdkafka: FAIL [thrd:locast:9092/bootstrap]: locast:9092/bootstrap: Failed to resolve 'locast:9092': Name or service not known (after 47ms in state CONNECT)
    at /root/.cargo/registry/src/rsproxy.cn-8f6827c7555bfaf8/rdkafka-0.28.0/src/client.rs:64

  2023-01-29T17:06:11.284 ERROR rdkafka::client: librdkafka: Global error: Resolve (Local: Host resolution failure): locast:9092/bootstrap: Failed to resolve 'locast:9092': Name or service not known (after 47ms in state CONNECT)
    at /root/.cargo/registry/src/rsproxy.cn-8f6827c7555bfaf8/rdkafka-0.28.0/src/client.rs:106

  2023-01-29T17:06:11.284 ERROR rdkafka::client: librdkafka: Global error: AllBrokersDown (Local: All broker connections are down): 1/1 brokers are down
    at /root/.cargo/registry/src/rsproxy.cn-8f6827c7555bfaf8/rdkafka-0.28.0/src/client.rs:106

Ok(())
Ok(())
  2023-01-29T17:06:13.237 ERROR librdkafka: librdkafka: FAIL [thrd:locast:9092/bootstrap]: locast:9092/bootstrap: Failed to resolve 'locast:9092': Name or service not known (after 0ms in state CONNECT, 1 identical error(s) suppressed)
    at /root/.cargo/registry/src/rsproxy.cn-8f6827c7555bfaf8/rdkafka-0.28.0/src/client.rs:64

  2023-01-29T17:06:13.237 ERROR rdkafka::client: librdkafka: Global error: Resolve (Local: Host resolution failure): locast:9092/bootstrap: Failed to resolve 'locast:9092': Name or service not known (after 0ms in state CONNECT, 1 identical error(s) suppressed)
    at /root/.cargo/registry/src/rsproxy.cn-8f6827c7555bfaf8/rdkafka-0.28.0/src/client.rs:106

Ok(())
Ok(())
Ok(())
Ok(())
Ok(())
Ok(())
  2023-01-29T17:06:20.244  WARN librdkafka: librdkafka: TERMINATE [thrd:app]: Producer terminating with 9 messages (81 bytes) still in queue or transit: use flush() to wait for outstanding message delivery
    at /root/.cargo/registry/src/rsproxy.cn-8f6827c7555bfaf8/rdkafka-0.28.0/src/client.rs:67

```

**Comment by hb0nes** [comment]
Just to add to this, some very basic idiomatic error handling is not possible with this library, unless I'm missing something, which I don't think I am:
```
    // 3 brokers are supplied, 2 of them are correct and 1 is purposely incorrect.
    let consumer: KafkaResult<StreamConsumer> = ClientConfig::new()
        .set("group.id", &group_id)
        .set("bootstrap.servers", &brokers)
        .set("enable.partition.eof", "false")
        .set("session.timeout.ms", "6000")
        .set("enable.auto.commit", "false")
        .create();
    let consumer = match consumer {
        Ok(v) => v,
        Err(why) => {
            dbg!(why);
            panic!();
        },
    };
    consumer
        .subscribe(&[&input_topic])
        .expect("Can't subscribe to specified topic");
```

I am trying to handle KafkaResult and are expecting an Err value.
However, when connecting to a broker that does not exist, librdkafka logs errors by itself:
```
 ERROR librdkafka      > librdkafka: FAIL [thrd:127.0.0.1:44908/bootstrap]: 127.0.0.1:44908/bootstrap: Connect to ipv4#127.0.0.1:44908 failed: Connection refused (after 0ms in state CONNECT)
 ERROR rdkafka::client > librdkafka: Global error: BrokerTransportFailure (Local: Broker transport failure): 127.0.0.1:44908/bootstrap: Connect to ipv4#127.0.0.1:44908 failed: Connection refused (after 0ms in state CONNECT)
```

KafkaResult is then Ok(), regardless of this error being detected/logged.

TL;DR: errors are being logged by the library, as opposed to feeding those errors back to the developer who can then choose to log them.

**Comment by wensiwei** [comment]
@davidblewett please see here ! help !

**Comment by theelderbeever** [comment]
For others in this thread. While not a great solution... I used some regex and the `dns_lookup` library to at least do an initial check on the broker address. Right now the following code just panics if the address can't be found but it is at least a little more useful than just having the error be logged "silently". This has helped me to crash the program on start up instead of not realizing that it is in an unhealthy state.

```rust
use dns_lookup::lookup_addr;
use dns_lookup::lookup_host;
use regex::Regex;

enum Addr {
    IpAddr(String),
    Hostname(String),
}

impl Addr {
    pub fn from(addr: &str) -> Self {
        if is_ipv4addr(addr) {
            Self::IpAddr(String::from(addr))
        } else {
            Self::Hostname(String::from(addr))
        }
    }
}

fn is_ipv4addr(addr: &str) -> bool {
    let re = Regex::new(
        r"^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$",
    )
    .unwrap();
    re.is_match(addr)
}

pub fn does_host_exist(addr: &str) {
    let parts = addr.split(':').collect::<Vec<&str>>();
    let addr = Addr::from(parts[0]);

    match addr {
        Addr::IpAddr(ip) => {
            let ip: std::net::IpAddr = ip.parse().unwrap();
            lookup_addr(&ip).expect("Host or Servname could not be found!");
        }
        Addr::Hostname(host) => {
            lookup_host(&host).expect("Host or Servname could not be found!");
        }
    }
}
```
